src/main/java/com/microsoft/sqlserver/jdbc/SQLServerResource.java

```
@@ -465,7 +465,16 @@ protected Object[][] getContents() {
         {"R_invalidSSLProtocol", "SSL Protocol {0} label is not valid. Only TLS, TLSv1, TLSv1.1, and TLSv1.2 are supported."},
         {"R_cancelQueryTimeoutPropertyDescription", "The number of seconds to wait to cancel sending a query timeout."},
         {"R_invalidCancelQueryTimeout", "The cancel timeout value {0} is not valid."},
-        {"R_useBulkCopyForBatchInsertPropertyDescription", "Whether the driver will use bulk copy API for batch insert operations"},
+        {"R_useBulkCopyForBatchInsertPropertyDescription", "Determines whether the driver will use bulk copy API for batch insert operations."},
+        {"R_bulkCopyOptionDefaultsBatchSizePropertyDescription", "The default batch size for bulk copy operations."},
+        {"R_bulkCopyOptionDefaultsTimeoutPropertyDescription", "The default timeout for bulk copy operations."},
+        {"R_bulkCopyOptionDefaultsCheckConstraintsPropertyDescription", "Determines whether to check constraints during bulk copy operations."},
+        {"R_bulkCopyOptionDefaultsFireTriggersPropertyDescription", "Determines whether to fire triggers during bulk copy operations."},
+        {"R_bulkCopyOptionDefaultsKeepIdentityPropertyDescription", "Determines whether to keep identity values during bulk copy operations."},
+        {"R_bulkCopyOptionDefaultsKeepNullsPropertyDescription", "Determines whether to keep null values during bulk copy operations."},
+        {"R_bulkCopyOptionDefaultsTableLockPropertyDescription", "Determines whether to use table lock during bulk copy operations."},
+        {"R_bulkCopyOptionDefaultsUseInternalTransactionPropertyDescription", "Determines whether to use internal transaction during bulk copy operations."},
+        {"R_bulkCopyOptionDefaultsAllowEncryptedValueModificationsPropertyDescription", "Determines whether to allow encrypted value modifications during bulk copy operations."},        
```

The descriptions here make me want to ensure we don't confuse users on what these settings apply to. As it is, it sounds to me like they would apply to the defaults when using the bulk copy APIs directly. E.g. Instantiating a new SQLServerBulkCopyOptions object outside of the context of a connection or not even passing options to a SQLServerBulkCopy object. See the many bulk copy API examples in the docs: https://learn.microsoft.com/en-us/sql/connect/jdbc/using-bulk-copy-with-the-jdbc-driver?view=sql-server-ver16#performing-a-dedicated-bulk-copy-operation-in-a-transaction for what I'm talking about where these new settings **_don't_** apply

I want to make sure it is clear that these options only apply within the scope of the useBulkCopyForBatchInsert option for PreparedStatements. (And without making the option name too long.)

I'm thinking along the lines of: bulkCopyForBatchInsertXXX
- bulkCopyForBatchInsertBatchSize
- bulkCopyForBatchInsertTimeout
- bulkCopyForBatchInsertFireTriggers
- etc...

Then these descriptions would become:

- The default batch size for bulk copy operations created from batch insert operations.
- The default timeout for bulk copy operations created from batch insert operations.
- Determines whether to check constraints during bulk copy operations created from batch insert operations.
- etc...

@lilgreenbird @tkyc - Let me know if these make more sense to you.

---

My understanding was the `default` in the name conveyed that these are connection defaults while one could use this connection for individual bulk copy operations with different option values.

---

yes agree this makes more sense, the "default" word is unnecessary and seems a bit misleading

---

> My understanding was the `default` in the name conveyed that these are connection defaults while one could use this connection for individual bulk copy operations with different option values.

Functionally, what does that look like? Meaning, given a set of code like the examples in the docs, what would the behavior change be?

The original request for this feature was to be able to control the bulk copy options used within the context of useBulkCopyForBatchInsert. Given that SQLServerBulkCopyOptions live outside of the context of a connection, I'm not sure it makes sense to try to apply the new options to contexts outside of useBulkCopyForBatchInsert. It might just cause confusion.

---

Agree, will make the changes.